sanic/headers.py

```diff
@@ -24,14 +24,132 @@ def parse_content_header(value: str) -> typing.Tuple[str, Options]:
     Mostly identical to cgi.parse_header and werkzeug.parse_options_header
     but runs faster and handles special characters better. Unescapes quotes.
     """
-    value = firefox_quote_escape.sub("%22", value)
+    value = _firefox_quote_escape.sub("%22", value)
     pos = value.find(";")
     if pos == -1:
         options = {}
     else:
         options = {
             m.group(1).lower(): m.group(2) or m.group(3).replace("%22", '"')
-            for m in parameter.finditer(value[pos:])
+            for m in _param.finditer(value[pos:])
         }
         value = value[:pos]
     return value.strip().lower(), options
+
+
+# https://tools.ietf.org/html/rfc7230#section-3.2.6 and
+# https://tools.ietf.org/html/rfc7239#section-4
+# This regex is for *reversed* strings because that works much faster for
+# right-to-left matching than the other way around. Be wary that all things are
+# a bit backwards! _rparam matches forwarded pairs alike ";key=value"
+_rparam = re.compile(f"(?:{_token}|{_quoted})={_token}\\s*($|[;,])", re.ASCII)
+
+
+def parse_forwarded(headers, config):
+    """Parse HTTP Forwarded headers.
+    Accepts only the last element with secret=`config.FORWARDED_SECRET`
+    :return: dict with matching keys (lowercase) and values, or None.
+    """
+    header = headers.getall("forwarded", None)
+    secret = config.FORWARDED_SECRET
+    if header is None or not secret:
+        return None
+    header = ",".join(header)  # Join multiple header lines
+    if secret not in header:
+        return None
+    # Loop over <separator><key>=<value> elements from right to left
+    ret = sep = pos = None
+    found = False
+    for m in _rparam.finditer(header[::-1]):
+        # Start of new element? (on parser skips and non-semicolon right sep)
+        if m.start() != pos or sep != ";":
+            if found:
+                return normalize(ret)
+            ret = {}
+        pos = m.end()
+        val_token, val_quoted, key, sep = m.groups()
+        key = key.lower()[::-1]
+        val = (val_token or val_quoted.replace('"\\', '"'))[::-1]
+        ret[key] = val
+        if key == "secret" and val == secret:
+            found = True
+        if found and sep != ";":
+            return normalize(ret)
+    return normalize(ret) if found else None
+
+
+def parse_xforwarded(headers, config):
+    """Parse X-Real-IP, X-Scheme and X-Forwarded-* headers."""
+    proxies_count = config.PROXIES_COUNT
+    if not proxies_count:
+        return None
+    h1, h2 = config.REAL_IP_HEADER, config.FORWARDED_FOR_HEADER
+    addr = h1 and headers.get(h1)
+    forwarded_for = h2 and headers.getall(h2, None)
+    if not addr and forwarded_for:
+        assert proxies_count == -1 or proxies_count > 0, config.PROXIES_COUNT
```

Comment:
Would it be a good idea to have a bit of a clearly defined error message with the assert ?

Or would it be easier to handle it the way the `addr` extraction is handled instead? Return `None` even if the `proxies_count` matching fails.